feat(core): add undecorated classes with decorated fields schematic

crisbeto opened the pull request:
Adds a schematic that adds a `Directive` decorator to undecorated classes that have fields that use Angular decorators.

alan-agius4 commented on packages/core/schematics/migrations/google3/undecoratedClassesWithDecoratedFieldsRule.ts:
```diff
@@ -0,0 +1,59 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Replacement, RuleFailure, Rules} from 'tslint';
```
As far as I recall using tslint for migrations is discouraged, and us writing migrations with tslint might give the impression that we encourage it's usage for migrations.

There are several issues when running migrations with tslint.
- it runs on files which are on disk and not using the schematic host, meaning that it might run on files which have changes not yet committed to disk.
- if the part of the schematic fails users will need to rollback the changes manually.

@clydin, can you confirm?

crisbeto replied:
These changes include both a regular schematic and a tslint rule which is used for G3 IIRC.

alan-agius4 replied:
Fair point

alan-agius4 commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/index.ts:
```diff
@@ -0,0 +1,90 @@
+  // if we run the migration for multiple tsconfig files which have intersecting
+  // source files, it can end up updating query definitions multiple times.
+  host.readFile = fileName => {
+    const buffer = tree.read(relative(basePath, fileName));
```
Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset which which breaks the CLI UpdateRecorder.

```
  host.readFile = fileName => {
    const buffer = tree.read(relative(basePath, fileName));
    // Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset which
    // which breaks the CLI UpdateRecorder.
    // See: https://github.com/angular/angular/pull/30719
    return buffer ? buffer.toString().replace(/^\uFEFF/, '') : undefined;
  };
```

alan-agius4 replied:
Probably we should create a helper for `createCompilerHost` to patches the `host.readFile` so that we don't need to do this in all migrations.

As otherwise, it will also be easy to forget to add it.

crisbeto replied:
I'll do it inline for now and we can make a pass over all the schematics afterwards.

devversion commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/README.md:
````diff
@@ -0,0 +1,23 @@
+```ts
+import { Input, Directive } from '@angular/core';
+
+@Directive({})
````
do we need the empty object literal here? AFAIK the design docs only mention that we add `@Directive()` (similar to what we did in the `undecorated-classes-with-di` migration)

crisbeto replied:
I'm still seeing it after rebasing:

<img width="677" alt="Screenshot at Aug 15 12-31-39" src="https://user-images.githubusercontent.com/4450522/63086519-b4a12e00-bf58-11e9-9656-b38c7d0e37d8.png">

devversion replied:
That's weird. Unless I miss something the merged commit https://github.com/angular/angular/commit/405515091040759c1e65e79b0dde3ab77e04975e definitely makes the `obj` param optional.

kara replied:
Seems like we should remove the object literal here.  When I check out your PR and remove the object, I don't get the type error on my end. Have you tried running tests and re-checking?

devversion commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/index.ts:
```diff
@@ -0,0 +1,90 @@
+  // We need to overwrite the host "readFile" method, as we want the TypeScript
+  // program to be based on the file contents in the virtual file tree. Otherwise
+  // if we run the migration for multiple tsconfig files which have intersecting
+  // source files, it can end up updating query definitions multiple times.
```
This seems like a copy-paste typo as `query definitions` are not part of this migration.

<sup>It looks like other migrations (`move-document`, `renderer-to-renderer2` and `injectable-pipe`) have the same issue.</sup>

devversion commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/index.ts:
```diff
@@ -0,0 +1,90 @@
+      }
+
+      // We don't need to go through the AST to insert the decorator, because the change
+      // is pretty basic. Also this has a better chance of preserving the user's formatting.
```
How does AST printing affect the user formatting here? If we would use AST and the printer, we'd still just prepend it to the class declaration (without affecting formatting).

In any case, it's totally fine to not use AST for this simple change. Just the comment might be confusing.

crisbeto replied:
What I meant with the comment is that if we printed out the entire class, the user formatting might get lost.

devversion commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/utils.ts:
```diff
@@ -0,0 +1,71 @@
+
+  sourceFile.forEachChild(function walk(node: ts.Node) {
+    if (ts.isClassDeclaration(node) &&
+        (!node.decorators || !getAngularDecorators(typeChecker, node.decorators).length)) {
```
An edge case here would be a class with `@Input`'s that has the `@Injectable()` decorator. This is surely an edge case.. but could be done by mistake. Should we be on the safe side and just check for either `@Directive` or `@Component`?

crisbeto replied:
This just checks whether there's _any_ Angular decorator so `@Injectable` should be caught as well.

devversion replied:
But isn't that the problem? Right now the class is just skipped if it has `@Injectable()` as the `getAngularDecorators(..).length` is not zero.

crisbeto replied:
I'm not sure how common this would be. If somebody tried doing it already, it's likely that something else broke for them down the line already.

devversion replied:
It's definitely not a common case. I don't think anything would break though if such a class just has `@Injectable()`. 

In any case it will break with Ivy if a directive extends from it. I don't feel strong about it :smile:

devversion commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/utils.ts:
```diff
@@ -0,0 +1,71 @@
+  const namedImports = getNamedImports(declaration);
+
+  if (namedImports) {
+    return !!namedImports.elements.find(element => {
```
You should be able to use `elements.some`?

crisbeto committed:
feat(core): add undecorated classes with decorated fields schematic

Adds a schematic that adds a `Directive` decorator to undecorated classes that have fields that use Angular decorators.

crisbeto commented:
Updated to address the feedback. Also no longer blocked.

kara commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/utils.ts:
```diff
@@ -0,0 +1,71 @@
+/** Name of the decorator that should be added to undecorated classes. */
+export const FALLBACK_DECORATOR = 'Directive';
+
+/** Finds all of the undercorated classes that have decorated fields within a file. */
```
```suggestion
/** Finds all of the undecorated classes that have decorated fields within a file. */
```

kara commented on packages/core/schematics/migrations/undecorated-classes-with-decorated-fields/index.ts:
```diff
@@ -0,0 +1,93 @@
+
+
+/**
+ * Migration that adds an Angular decorator to classes that have Angular decorator field decorators.
```
```suggestion
 * Migration that adds an Angular decorator to classes that have Angular field decorators.
```

crisbeto commented:
The feedback has been addressed @kara.

kara commented on packages/core/schematics/migrations/google3/undecoratedClassesWithDecoratedFieldsRule.ts:
```diff
@@ -0,0 +1,58 @@
+
+
+/**
+ * TSLint rule that adds an Angular decorator to classes that have Angular decorator field
```
```suggestion
 * TSLint rule that adds an Angular decorator to classes that have Angular field
```

kara reviewed (approved):
LGTM, one more comment to fix though (and CI looks unhappy?)

crisbeto committed:
fixup! feat(core): add undecorated classes with decorated fields schematic

crisbeto commented:
Addressed the last set of feedback and pushed something to fix the CI.

kara commented:
[VE presubmit](https://test.corp.google.com/ui#id=OCL:264759030:BASE:264759037:1566448607817:52b4712f)

[Ivy presubmit](http://test/OCL:264759030:BASE:264759764:1566449052443:62ce0328)
